compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -867,6 +887,28 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> std::fmt::Debug for DumpAllocs<'a,
             write!(fmt, "{}", display_allocation(tcx, alloc))
         }
 
+        // Format the backtrace, imitating the RUST_BACKTRACE panic formatting.
+        let write_backtrace = |fmt: &mut std::fmt::Formatter<'_>,
```

Comment:
This could dispatch to a machine hook and then use the Miri backtrace printing functionality.

Reply:
The Miri backtrace functionality eventually ends with `DiagnosticBuilder::emit`, but the alloc dumping logic that I want to plug into the middle of here is used in both `DumpAllocs` and `PlacePrinter`, and in those places it is used through `Debug`, so I need to direct the backtrace to a `std::fmt::Formatter`, and I can't figure out how to do that with a `DiagnosticBuilder`.

Reply:
Oh you are entering the fun territory. If you want to try it out:

`HumanReadableErrorType::new` can be created with its `dst` argument being a `Box<YourThing>` where `YourThing` implements `Write` and contains an `Arc<Mutex<Vec<u8>>>`. Then you can use `emit_diagnostic` to have your diagnostic written to that buffer. You need to keep a clone of the `Arc` around, but you can then read the bytes out of it and render them as a string to anything you want.

This is a bit whacky, but it does the job somewhat. We may want to see if we can rewrite everything to use the main diagnostic renderer instead of writing to stderr directly, this way all the json diagnostics processing logic will extract the rendered part instead of us patching around it.

Reply:
Yeah I think I'm going to put the machine hooks at a higher level and punt the entire leak report to Miri, and just share the leak finding code.